keps/sig-network/1880-multiple-service-cidrs/README.md

```diff
@@ -0,0 +1,954 @@
+tests].**
+
+[conformance tests]:
+    https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md
+
+#### Deprecation
+
+### Upgrade / Downgrade / Version Skew Strategy
+
+The source of truth are the IPs assigned to the Services, both the old and new methods have
+reconcile loops that rebuild the state of the allocators based on the assigned IPs to the Services,
+this allows to support upgrades and skewed clusters.
+
+Clusters running with the new allocation model will not keep running the reconcile loops that
+keep the bitmap used by the allocators.
+
+Since the new allocation model will remove some of the limitations of the current model, skewed
+versions and downgrades can only work if the configurations are fully compatible, per example,
+current CIDRs are limited to a /112 max for IPv6, if an user configures a /64 to their IPv6 subnets
+in the new model, and IPs are assigned out of the first /112 block, the old allocator based in
+bitmap will not be able to use those IPs creating an inconsistency in the cluster.
+
+It will be required that those Services are recreated to get IP addresses inside the configured
+ranges, for consistency, but there should not be any functional problem in the cluster if the
+Service implementation (kube-proxy, ...) is able to handle those IPs.
+
+Example:
+
+- flags set to 10.0.0.0/20
+- upgrade to 1.25 with alpha gate
+- apiserver create a default ServiceCIDRConfig object for 10.0.0.0/20
+- user creates a new ServiceCIDR for 192.168.1.0/24
+- create a Service which gets 192.168.1.1
+- rollback or disable the gate
+- the apiserver repair loops will generate periodic events informing the user that the Service with the IP allocated
+  is not within the configured range
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: ServiceCIDRConfig
+  - Components depending on the feature gate: kube-apiserver, kube-controller-manager
+
+###### Does enabling the feature change any default behavior?
+
+The time to reuse an already allocated ClusterIP to a Service will be longer, since the ClusterIPs
+now depend on an IPAddress object that is protected by a finalizer.
+
+The bootstrap logic has been updated to deal with the problem of multiple apiservers with different
+configurations, making it more flexible and resilient.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+If the feature is disabled the old allocator logic will reconcile all the current allocated IPs
+based on the Services existing on the cluster.
+
+If there are IPAddresses allocated outside of the configured Service IP Ranges in the apiserver flags,
+the apiserver will generate events referencing the Services using IPs outside of the configured range.
```

Comment:
Will the services function? I suspect not...

If this breaks the services anyway, perhaps we're better off doing some automated recreation? Or maybe better, re-allocation of the cluster IPs? Because recreation will possibly lose LB IPs...

This is pretty brutal rollback behavior.

Reply:
no, it will work - same as today.  Consumers assume that if an IP made its way into a Service they should use that IP.  The repair loop looks at it ind says "that's weird" but it can't really do much about it - the IP may have been statically configured.  It's the same as if I changed my apiserver flags to a larter range, used it, and then changed it back.

Reply:
You have cluster with service-cidr 192.168.1.0/24 and 10.0.0.0/24
When you roll back you can have only one, let's say 192.168.1.0/24
The Services with 10.0.0.0/24 addresses will keep working, you'll receive the events that they don't belong to the defined range.

ClusterIP are inmutable, you can only recreate the Service

Reply:
heh https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/2450-Remove-knowledge-of-pod-cluster-CIDR-from-iptables-rules/README.md :smile: